core/trino-spi/src/main/java/io/trino/spi/block/FixedBinaryBlockEncoding.java

```diff
@@ -0,0 +1,98 @@
+
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+
+import static io.trino.spi.block.EncoderUtil.decodeNullBits;
+import static io.trino.spi.block.EncoderUtil.encodeNullsAsBits;
+
+public class FixedBinaryBlockEncoding
+        implements BlockEncoding
+{
+    public static final String NAME = "FIXED_BINARY";
+
+    @Override
+    public String getName()
+    {
+        return NAME;
+    }
+
+    @Override
+    public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceOutput, Block block)
+    {
+        int positionCount = block.getPositionCount();
+        sliceOutput.appendInt(positionCount);
+        FixedBinaryBlockOutput blockOutput = (FixedBinaryBlockOutput) block;
+        int positionWidth = blockOutput.getPositionWidth();
+        sliceOutput.appendInt(positionWidth);
+
+        encodeNullsAsBits(sliceOutput, block);
+
+        if (!block.mayHaveNull()) {
+            sliceOutput.writeBytes(blockOutput.getValuesSlice());
+        }
+        else {
+            long[] valuesWithoutNull = new long[positionCount * positionWidth];
+            int nonNullPositionCount = 0;
+            for (int i = 0; i < positionCount; i++) {
+                for (int j = 0; j < positionWidth; j++) {
+                    valuesWithoutNull[nonNullPositionCount + j] = block.getLong(i, Long.BYTES * j);
+                }
+                int isNotNull = !block.isNull(i) ? 1 : 0;
+                nonNullPositionCount += positionWidth * isNotNull;
+            }
+
+            sliceOutput.writeInt(nonNullPositionCount / positionWidth);
+            sliceOutput.writeBytes(Slices.wrappedLongArray(valuesWithoutNull, 0, nonNullPositionCount));
+        }
+    }
+
+    @Override
+    public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceInput)
+    {
+        int positionCount = sliceInput.readInt();
+        int positionWidth = sliceInput.readInt();
+
+        boolean[] valueIsNull = decodeNullBits(sliceInput, positionCount).orElse(null);
+
+        long[] values = new long[positionCount * positionWidth];
+        if (valueIsNull == null) {
+            sliceInput.readBytes(Slices.wrappedLongArray(values));
+        }
+        else {
+            int nonNullPositionCount = sliceInput.readInt();
+            sliceInput.readBytes(Slices.wrappedLongArray(values, 0, nonNullPositionCount * positionWidth));
+            int position = positionWidth * (nonNullPositionCount - 1);
+            for (int i = positionCount - 1; i >= 0 && position >= 0; i--) {
+                System.arraycopy(values, position, values, positionWidth * i, positionWidth);
+                if (!valueIsNull[i]) {
+                    position -= positionWidth;
+                }
+            }
+        }
+
+        return new FixedBinaryBlock(0, positionCount, valueIsNull, values, positionWidth * Long.BYTES);
+    }
+
+    public interface FixedBinaryBlockOutput
```

Comment:
Use similar pattern as in `io.trino.spi.block.LongArrayBlockEncoding#getValuesSlice`.
I would rather keep it consistent (for now) or change it for other blocks too (later).

This interface is artificial anyway with two arbitrary methods.
At some point we will most likely refactor blocks anyway so that it's possible to operate on raw arrays

Reply:
this was removed in the following commit (i.e. commits are not squashed yet)